### ISSUE TITLE
Incorrect understanding of implicit namespace packages

### ISSUE THREAD

**Issue by AlexWaygood** [microsoft/pyright#8408]
(I discovered this bug because I was looking for edge cases for Ruff's work-in-progress multifile module resolver. I noticed that the same bug existed in pyright, and thought you might be interested in knowing about it.)

**Describe the bug**
Consider the following directory structure:

- foo/
  - foo/bar.py
- foo.py
- script.py

Where `foo.py` is empty, and the contents of `foo/bar.py` are:

```py
Y = 42
```

The contents of `script.py` are:

```py
# pyright: strict

import foo.bar

print(foo.bar.Y)
```

At runtime, running `python script.py` results in the following exception, as Python does not consider there to be an importable `foo.bar` module:

```pytb
Traceback (most recent call last):
  File "/Users/alexw/dev/experiment/script.py", line 3, in <module>
    import foo.bar
ModuleNotFoundError: No module named 'foo.bar'; 'foo' is not a package
```

This is because the single-file module in the root directory takes precedence over the implicit namespace `foo` package in the same directory. (If `foo/__init__.py` existed, then the `foo` directory would be a regular package that _would_ take precedence over the single-file `foo` module -- but `foo/__init__.py` does not exist).

Pyright, however, does not detect this; it thinks there are no issues in `script.py`.

**Comment by erictraut** [comment]
Thanks for the bug report — and for the clarification. Import resolution in Python sure is complicated.

It appears that mypy also has this bug, although I know that it has many issues with namespace package resolution.

This will be addressed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.372.
